Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp

```diff
@@ -343,20 +343,51 @@ void RemoteRenderingBackend::destroyGetPixelBufferSharedMemory()
     m_getPixelBufferSharedMemory = nullptr;
 }
 
-
-void RemoteRenderingBackend::cacheNativeImage(ShareableBitmap::Handle&& handle, RenderingResourceIdentifier nativeImageIdentifier)
+void RemoteRenderingBackend::nativeImageBitmap(RenderingResourceIdentifier imageIdentifier, CompletionHandler<void(std::optional<ShareableBitmap::Handle>)>&& completionHandler)
+{
+    std::optional<ShareableBitmap::Handle> result;
+    [&] {
+        RefPtr image = m_remoteResourceCache.cachedNativeImage(imageIdentifier);
+        MESSAGE_CHECK(image, "NativeImage not cached.");
+        RefPtr<ShareableBitmap> bitmap;
+#if USE(CG)
+        bitmap = ShareableBitmap::createFromImagePixels(*image);
+#endif
+        if (!bitmap)
+            bitmap = ShareableBitmap::createFromImageDraw(*image, DestinationColorSpace { image->colorSpace() });
+        if (!bitmap)
+            return;
+        auto handle = bitmap->createHandle();
+        if (!handle)
+            return;
+        if (sharedResourceCache().resourceOwner())
+            handle->setOwnershipOfMemory(sharedResourceCache().resourceOwner(), WebCore::MemoryLedger::Graphics);
+        auto platformImage = bitmap->createPlatformImage();
+        if (!platformImage) {
+            ASSERT_NOT_REACHED();
+            return;
+        }
+        image->replacePlatformImage(WTFMove(platformImage));
+        result = WTFMove(handle);
+    }();
+    ASSERT(result);
+    completionHandler(WTFMove(result));
+}
+
+void RemoteRenderingBackend::cacheNativeImage(ShareableBitmap::Handle&& handle, RenderingResourceIdentifier imageIdentifier)
 {
     ASSERT(!RunLoop::isMain());
 
     auto bitmap = ShareableBitmap::create(WTFMove(handle));
     if (!bitmap)
         return;
 
-    auto image = NativeImage::create(bitmap->createPlatformImage(DontCopyBackingStore, ShouldInterpolate::Yes), nativeImageIdentifier);
+    auto image = NativeImage::create(bitmap->createPlatformImage(DontCopyBackingStore, ShouldInterpolate::Yes));
```

Comment:
Why did you remove the passing of the identifier here?

Reply:
> In here it's only done half-away, e.g. the code is partially adjusted. NativeImages still hold identifiers, since SourceImage accesses it directly and incorrectly. Later on, will be fixed and the line above is already done.

Right, of course, I should have remembered that you'd already done this for other types.

It seemed a bit strange for the identifier keying the hash map to not match the identifier of the object, even though it was unused. Makes sense in the context of the larger plan though!